File: src/simulation/simulation.ts

type Resource = {
    id: string;
    name: string;
    resourceType: string;
};

class MissingResource {
    constructor(public name: string, public resourceType: string) {}

    canBeAllocatedBy(resource: Resource): boolean {
        return this.resourceType === resource.resourceType;
    }
}

class Project {
    constructor(
        public name: string,
        public estimatedCost: number,
        public estimatedEarnings: number,
        public risk: number,
        public penalty: number,
        public missingResource: MissingResource[]
    ) {}

    estimatedProfit(): number {
        return this.estimatedEarnings - this.estimatedCost;
    }
}

type CalculateProfitQuery = {
    projects: Project[];
    availableResources: Resource[];

    orderedProjects: () => Project[];
};

type Result = {
    profit: number;
    projects: Project[];

    toString: () => string;
};

class ChoseOptimalProjects {
    apply(query: CalculateProfitQuery): Result {
        const totalResources = query.availableResources.length;
        let dp: number[] = new Array(totalResources + 1).fill(0);
        let projectLists: Project[][] = new Array(totalResources + 1).fill([]);
        let allocatedResources: Set<Resource>[] = new Array(totalResources + 1).fill(new Set());

        for (let project of query.orderedProjects()) {
            let allocatableResources = this.resourcesFromRequired(project.missingResource, query.availableResources);
            if (allocatableResources.length === 0) continue;
            let projectProfit = project.estimatedProfit();
            let allocatableResourcesCount = allocatableResources.length;

            for (let j = totalResources; j >= allocatableResourcesCount; j--) {
                if (!this.isResourceAllocated(allocatableResources, allocatedResources[j - allocatableResourcesCount])) {
                    if (dp[j] < projectProfit + dp[j - allocatableResourcesCount]) {
                        dp[j] = projectProfit + dp[j - allocatableResourcesCount];
                        projectLists[j] = [...projectLists[j - allocatableResourcesCount], project];
                        allocatedResources[j] = new Set([...allocatedResources[j], ...allocatableResources]);
                    }
                }
            }
        }
        return {
            profit: dp[totalResources],
            projects: projectLists[totalResources],
            toString: () => `Result{profit=${dp[totalResources]}, projects=${projectLists[totalResources]}}`
        };
    }

    private isResourceAllocated(required: Resource[], allocated: Set<Resource>): boolean {
        return required.some(res => allocated.has(res));
    }

    private resourcesFromRequired(requiredResources: MissingResource[], availableResources: Resource[]): Resource[] {
        return requiredResources.flatMap(req => availableResources.filter(resource => resource.name === req.name && resource.resourceType === req.resourceType));
    }
}
function test() {
    // Creating resources
    const skill1: Resource = { id: "Ania", name: "Java Developer", resourceType: "SKILL" };
    const skill2: Resource = { id: "Marek", name: "Web Designer", resourceType: "SKILL" };
    const tool1: Resource = { id: "XPS", name: "Laptop", resourceType: "DEVICE" };

    // Creating projects
    const project1 = new Project(
        "Website Creation",
        1000,
        3000,
        30,
        500,
        [new MissingResource("Web Designer", "SKILL")]
    );

    const project2 = new Project(
        "Database Setup",
        1500,
        4000,
        50,
        800,
        [new MissingResource("Web Designer", "SKILL")]
    );

    const projectsToOptimize: Project[] = [project1, project2];
    const resourcesWeHave: Resource[] = [skill1, skill2, tool1];
    const query: CalculateProfitQuery = { projects: projectsToOptimize, availableResources: resourcesWeHave, orderedProjects: function () { return this.projects.sort((a, b) => b.estimatedProfit() - a.estimatedProfit()); } };
    const chooser = new ChoseOptimalProjects();
    const result = chooser.apply(query);

    console.log("Max Profit:", result.profit);
    console.log("Selected Projects:", result.projects);
}

test();

File: src/optimization/simulation.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Simulation
{
    public record Resource(string Id, string Name, string ResourceType);

    public record MissingResource(string Name, string ResourceType)
    {
        public bool CanBeAllocatedBy(Resource resource)
        {
            return this.ResourceType == resource.ResourceType;
        }
    }

    public record Project(string Name, double EstimatedCost, double EstimatedEarnings, int Risk, double Penalty, List<MissingResource> MissingResources)
    {
        public double EstimatedProfit()
        {
            return EstimatedEarnings - EstimatedCost;
        }

		public override string ToString()
        {
            return $"{Name}";
        }
    }

    public class ChoseOptimalProjects
    {
        public Result Apply(CalculateProfitQuery query)
        {
            int totalResources = query.AvailableResources.Count;
            double[] dp = new double[totalResources + 1];
            List<Project>[] projectLists = new List<Project>[totalResources + 1];
            List<HashSet<Resource>> allocatedResources = new List<HashSet<Resource>>(totalResources + 1);

            for (int i = 0; i <= totalResources; i++)
            {
                projectLists[i] = new List<Project>();
                allocatedResources.Add(new HashSet<Resource>());
            }

            foreach (Project project in query.OrderedProjects())
            {
                List<Resource> allocatableResources = ResourcesFromRequired(project.MissingResources, query.AvailableResources);

                if (!allocatableResources.Any())
                    continue;

                double projectProfit = project.EstimatedProfit();
                int allocatableResourcesCount = allocatableResources.Count;

                for (int j = totalResources; j >= allocatableResourcesCount; j--)
                {
                    if (!IsResourceAllocated(allocatableResources, allocatedResources[j - allocatableResourcesCount]))
                    {
                        if (dp[j] < projectProfit + dp[j - allocatableResourcesCount])
                        {
                            dp[j] = projectProfit + dp[j - allocatableResourcesCount];

                            projectLists[j] = new List<Project>(projectLists[j - allocatableResourcesCount]);
                            projectLists[j].Add(project);

                            allocatedResources[j].UnionWith(allocatableResources);
                        }
                    }
                }
            }
            return new Result(dp[totalResources], projectLists[totalResources]);
        }

        private bool IsResourceAllocated(List<Resource> required, HashSet<Resource> allocated)
        {
            return required.Any(allocated.Contains);
        }

        private List<Resource> ResourcesFromRequired(List<MissingResource> requiredResources, List<Resource> availableResources)
        {
            return requiredResources.SelectMany(req => availableResources.Where(resource => resource.Name == req.Name && resource.ResourceType == req.ResourceType)).ToList();
        }
    }

    public record CalculateProfitQuery(List<Project> Projects, List<Resource> AvailableResources)
    {
        public List<Project> OrderedProjects()
        {
            return Projects.OrderByDescending(p => p.EstimatedProfit()).ToList();
        }
    }

    public record Result(double Profit, List<Project> Projects)
    {
        public override string ToString()
        {
            return $"Result {{ profit = {Profit}, projects = {Projects} }}";
        }
    }

    public class Test
    {
        public static void Main()
        {
            Resource skill1 = new Resource("Ania", "Java Developer", "SKILL");
            Resource skill2 = new Resource("Marek", "Web Designer", "SKILL");
            Resource tool1 = new Resource("XPS", "Laptop", "DEVICE");

            Project project1 = new Project("Website Creation", 1000, 3000, 30, 500, new List<MissingResource> { new MissingResource("Web Designer", "SKILL") });
            Project project2 = new Project("Database Setup", 1500, 4000, 50, 800, new List<MissingResource> { new MissingResource("Web Designer", "SKILL") });

            List<Project> projectsToOptimize = new List<Project> { project1, project2 };
            List<Resource> resourcesWeHave = new List<Resource> { skill1, skill2, tool1 };
            CalculateProfitQuery query = new CalculateProfitQuery(projectsToOptimize, resourcesWeHave);
            Result result = new ChoseOptimalProjects().Apply(query);

            Console.WriteLine("Max Profit: " + result.Profit);
			Console.WriteLine("Selected Projects:");
            result.Projects.ForEach(project => Console.WriteLine(project.ToString()));
        }
    }
}
